File: Python/search_2dMatrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        t_r, b_r = 0, len(matrix) -1
        l_c, r_c = 0, len(matrix[0]) -1
        while t_r <= b_r:
            m_r = (t_r + b_r)//2
            if matrix[m_r][-1] < target:
                    t_r = m_r + 1
            elif matrix[m_r][0] > target:
                    b_r = m_r -1
            else:
                break
        if not (t_r <= b_r):
            return False

        while l_c <= r_c:
            m_c = (l_c+r_c)//2
            if matrix[m_r][m_c] < target:
                l_c = m_c + 1
            elif matrix[m_r][m_c] > target:
                r_c = m_c-1
            else:
                return True
        return False

File: Cpp/search_2dMatrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int left, top;
        left = top = 0;
        int bot = matrix.size() - 1;
        int right = matrix[0].size() - 1;

        while(top <= bot){
            int mid_row = (top + bot) / 2;
            cout << top << "   " << bot << "   "  << mid_row << endl;
            if (target > matrix[mid_row][right]) top = mid_row + 1;
            else if(target < matrix[mid_row][0]) bot = mid_row - 1;
            else {
                // cout << top << " break " << bot << "   "  << mid_row << endl;
                break; 
            }         
        }

        if (top > bot) return false;
        int mid_row = (top + bot) / 2;

        while(left <= right)
        {
            int mid_col = (left + right) / 2;
            // cout << matrix[mid_row][mid_col] << "   " << matrix[mid_row][0] <<endl;
            if (target > matrix[mid_row][mid_col]) left = mid_col + 1;
            else if(target < matrix[mid_row][mid_col]) right = mid_col - 1;
            else return true;           
        }
        return false;
    }
    
};